File: blaise-graphics/src/main/java/com/googlecode/blaisemath/graphics/swing/JGraphics.java
package com.googlecode.blaisemath.graphics.swing;

/*
 * #%L
 * BlaiseGraphics
 * --
 * Copyright (C) 2009 - 2020 Elisha Peterson
 * --
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */


import com.googlecode.blaisemath.graphics.swing.render.PathRenderer;
import com.googlecode.blaisemath.graphics.swing.render.ImageRenderer;
import com.googlecode.blaisemath.graphics.swing.render.ShapeRenderer;
import com.googlecode.blaisemath.graphics.swing.render.MarkerRenderer;
import com.googlecode.blaisemath.graphics.swing.render.IconRenderer;
import com.googlecode.blaisemath.graphics.swing.render.TextRenderer;
import com.google.common.graph.EndpointPair;
import com.googlecode.blaisemath.graphics.impl.DelegatingNodeLinkGraphic;
import com.googlecode.blaisemath.graphics.DelegatingPrimitiveGraphic;
import com.googlecode.blaisemath.graphics.PrimitiveGraphic;
import com.googlecode.blaisemath.style.AttributeSet;
import com.googlecode.blaisemath.style.ObjectStyler;
import com.googlecode.blaisemath.style.Styles;
import com.googlecode.blaisemath.coordinate.OrientedPoint2D;
import com.googlecode.blaisemath.primitive.AnchoredIcon;
import com.googlecode.blaisemath.primitive.AnchoredImage;
import com.googlecode.blaisemath.primitive.AnchoredText;
import java.awt.AlphaComposite;
import java.awt.BasicStroke;
import java.awt.Composite;
import java.awt.Graphics2D;
import java.awt.Image;
import java.awt.Shape;
import java.awt.geom.Point2D;
import javax.swing.Icon;

/**
 * Factory methods for creating basic java2d-based graphics.
 * 
 * @author Elisha Peterson
 */
public class JGraphics {
    
    /** Default stroke of 1 unit width. */
    public static final BasicStroke DEFAULT_STROKE = new BasicStroke(1.0f);
    /** Default composite */
    public static final Composite DEFAULT_COMPOSITE = AlphaComposite.getInstance(AlphaComposite.SRC_OVER);

    // utility class
    private JGraphics() {
    }
    
    //region FACTORIES
    
    public static PrimitiveGraphic<Shape, Graphics2D> path(Shape primitive) {
        return new PrimitiveGraphic<>(primitive, Styles.DEFAULT_PATH_STYLE.copy(), PathRenderer.getInstance());
    }

    public static PrimitiveGraphic<Shape, Graphics2D> path(Shape primitive, AttributeSet style) {
        return new PrimitiveGraphic<>(primitive, style, PathRenderer.getInstance());
    }

    public static <S> DelegatingPrimitiveGraphic<S, Shape, Graphics2D> path(S source, Shape primitive, ObjectStyler<S> styler) {
        return new DelegatingPrimitiveGraphic<>(source, primitive, styler, PathRenderer.getInstance());
    }

    public static PrimitiveGraphic<Shape, Graphics2D> shape(Shape primitive) {
        return new PrimitiveGraphic<>(primitive, Styles.DEFAULT_SHAPE_STYLE.copy(), ShapeRenderer.getInstance());
    }

    public static PrimitiveGraphic<Shape, Graphics2D> shape(Shape primitive, AttributeSet style) {
        return new PrimitiveGraphic<>(primitive, style, ShapeRenderer.getInstance());
    }

    public static <S> DelegatingPrimitiveGraphic<S, Shape, Graphics2D> shape(S source, Shape primitive, ObjectStyler<S> styler) {
        return new DelegatingPrimitiveGraphic<>(source, primitive, styler, ShapeRenderer.getInstance());
    }

    public static PrimitiveGraphic<Point2D, Graphics2D> point(Point2D primitive) {
        return new PrimitiveGraphic<>(primitive, Styles.DEFAULT_POINT_STYLE.copy(), MarkerRenderer.getInstance());
    }

    public static PrimitiveGraphic<Point2D, Graphics2D> point(Point2D primitive, AttributeSet style) {
        return new PrimitiveGraphic<>(primitive, style, MarkerRenderer.getInstance());
    }

    public static <S> DelegatingPrimitiveGraphic<S, Point2D, Graphics2D> point(S source, Point2D primitive, ObjectStyler<S> styler) {
        return new DelegatingPrimitiveGraphic<>(source, primitive, styler, MarkerRenderer.getInstance());
    }

    public static PrimitiveGraphic<Point2D, Graphics2D> marker(OrientedPoint2D primitive) {
        return new PrimitiveGraphic<>(primitive, Styles.DEFAULT_POINT_STYLE.copy(), MarkerRenderer.getInstance());
    }

    public static PrimitiveGraphic<Point2D, Graphics2D> marker(OrientedPoint2D primitive, AttributeSet style) {
        return new PrimitiveGraphic<>(primitive, style, MarkerRenderer.getInstance());
    }
    
    public static <S> DelegatingPrimitiveGraphic<S, Point2D, Graphics2D> marker(S source, OrientedPoint2D primitive, ObjectStyler<S> styler) {
        return new DelegatingPrimitiveGraphic<>(source, primitive, styler, MarkerRenderer.getInstance());
    }

    public static PrimitiveGraphic<AnchoredText, Graphics2D> text(AnchoredText primitive) {
        return new PrimitiveGraphic<>(primitive, Styles.DEFAULT_TEXT_STYLE.copy(), TextRenderer.getInstance());
    }

    public static PrimitiveGraphic<AnchoredText, Graphics2D> text(AnchoredText primitive, AttributeSet style) {
        return new PrimitiveGraphic<>(primitive, style, TextRenderer.getInstance());
    }

    public static <S> DelegatingPrimitiveGraphic<S, AnchoredText, Graphics2D> text(S source, AnchoredText primitive, ObjectStyler<S> styler) {
        return new DelegatingPrimitiveGraphic<>(source, primitive, styler, TextRenderer.getInstance());
    }

    public static PrimitiveGraphic<AnchoredImage, Graphics2D> image(AnchoredImage primitive) {
        return new PrimitiveGraphic<>(primitive, AttributeSet.EMPTY, ImageRenderer.getInstance());
    }

    public static PrimitiveGraphic<AnchoredImage, Graphics2D> image(double x, double y, double wid, double ht, Image image, String ref) {
        return new PrimitiveGraphic<>(new AnchoredImage(x, y, wid, ht, image, ref), AttributeSet.EMPTY, ImageRenderer.getInstance());
    }

    public static PrimitiveGraphic<AnchoredIcon, Graphics2D> icon(AnchoredIcon icon) {
        return new PrimitiveGraphic<>(icon, AttributeSet.EMPTY, IconRenderer.getInstance());
    }

    public static PrimitiveGraphic<AnchoredIcon, Graphics2D> icon(Icon icon, double x, double y) {
        return new PrimitiveGraphic<>(new AnchoredIcon(x, y, icon), AttributeSet.EMPTY, IconRenderer.getInstance());
    }

    public static <S> DelegatingNodeLinkGraphic<S, EndpointPair<S>, Graphics2D> nodeLink() {
        return new DelegatingNodeLinkGraphic<>(MarkerRenderer.getInstance(), TextRenderer.getInstance(), PathRenderer.getInstance());
    }
    
    //endregion
    
}

File: blaise-graphics/src/main/java/com/googlecode/blaisemath/graphics/swing/JGraphics.kt
package com.googlecode.blaisemath.graphics.swing

import com.google.common.graph.EndpointPair
import com.googlecode.blaisemath.coordinate.OrientedPoint2D
import com.googlecode.blaisemath.graphics.DelegatingPrimitiveGraphic
import com.googlecode.blaisemath.graphics.PrimitiveGraphic
import com.googlecode.blaisemath.graphics.impl.DelegatingNodeLinkGraphic
import com.googlecode.blaisemath.graphics.svg.SvgPathCoderTest
import com.googlecode.blaisemath.graphics.swing.ClippedImageTestFrame
import com.googlecode.blaisemath.graphics.swing.ContextMenuTestFrame
import com.googlecode.blaisemath.graphics.swing.HelloWorldTestFrame
import com.googlecode.blaisemath.graphics.swing.PanAndZoomTestFrame
import com.googlecode.blaisemath.graphics.swing.SelectionTestFrame
import com.googlecode.blaisemath.graphics.swing.TooltipTestFrame
import com.googlecode.blaisemath.graphics.swing.render.*
import com.googlecode.blaisemath.primitive.AnchoredIcon
import com.googlecode.blaisemath.primitive.AnchoredImage
import com.googlecode.blaisemath.primitive.AnchoredText
import com.googlecode.blaisemath.style.AttributeSet
import com.googlecode.blaisemath.style.ObjectStyler
import com.googlecode.blaisemath.style.Styles
import junit.framework.TestCase
import java.awt.*
import java.awt.geom.Point2D
import javax.swing.Icon

/*
* #%L
* BlaiseGraphics
* --
* Copyright (C) 2009 - 2021 Elisha Peterson
* --
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*      http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
* #L%
*/ /**
 * Factory methods for creating basic java2d-based graphics.
 *
 * @author Elisha Peterson
 */
object JGraphics {
    /** Default stroke of 1 unit width.  */
    val DEFAULT_STROKE: BasicStroke? = BasicStroke(1.0f)

    /** Default composite  */
    val DEFAULT_COMPOSITE: Composite? = AlphaComposite.getInstance(AlphaComposite.SRC_OVER)

    //region FACTORIES
    fun path(primitive: Shape?): PrimitiveGraphic<Shape?, Graphics2D?>? {
        return PrimitiveGraphic(primitive, Styles.DEFAULT_PATH_STYLE.copy(), PathRenderer.Companion.getInstance())
    }

    fun path(primitive: Shape?, style: AttributeSet?): PrimitiveGraphic<Shape?, Graphics2D?>? {
        return PrimitiveGraphic(primitive, style, PathRenderer.Companion.getInstance())
    }

    fun <S> path(source: S?, primitive: Shape?, styler: ObjectStyler<S?>?): DelegatingPrimitiveGraphic<S?, Shape?, Graphics2D?>? {
        return DelegatingPrimitiveGraphic(source, primitive, styler, PathRenderer.Companion.getInstance())
    }

    fun shape(primitive: Shape?): PrimitiveGraphic<Shape?, Graphics2D?>? {
        return PrimitiveGraphic(primitive, Styles.DEFAULT_SHAPE_STYLE.copy(), ShapeRenderer.Companion.getInstance())
    }

    fun shape(primitive: Shape?, style: AttributeSet?): PrimitiveGraphic<Shape?, Graphics2D?>? {
        return PrimitiveGraphic(primitive, style, ShapeRenderer.Companion.getInstance())
    }

    fun <S> shape(source: S?, primitive: Shape?, styler: ObjectStyler<S?>?): DelegatingPrimitiveGraphic<S?, Shape?, Graphics2D?>? {
        return DelegatingPrimitiveGraphic(source, primitive, styler, ShapeRenderer.Companion.getInstance())
    }

    fun point(primitive: Point2D?): PrimitiveGraphic<Point2D?, Graphics2D?>? {
        return PrimitiveGraphic(primitive, Styles.DEFAULT_POINT_STYLE.copy(), MarkerRenderer.Companion.getInstance())
    }

    fun point(primitive: Point2D?, style: AttributeSet?): PrimitiveGraphic<Point2D?, Graphics2D?>? {
        return PrimitiveGraphic(primitive, style, MarkerRenderer.Companion.getInstance())
    }

    fun <S> point(source: S?, primitive: Point2D?, styler: ObjectStyler<S?>?): DelegatingPrimitiveGraphic<S?, Point2D?, Graphics2D?>? {
        return DelegatingPrimitiveGraphic(source, primitive, styler, MarkerRenderer.Companion.getInstance())
    }

    fun marker(primitive: OrientedPoint2D?): PrimitiveGraphic<Point2D?, Graphics2D?>? {
        return PrimitiveGraphic(primitive, Styles.DEFAULT_POINT_STYLE.copy(), MarkerRenderer.Companion.getInstance())
    }

    fun marker(primitive: OrientedPoint2D?, style: AttributeSet?): PrimitiveGraphic<Point2D?, Graphics2D?>? {
        return PrimitiveGraphic(primitive, style, MarkerRenderer.Companion.getInstance())
    }

    fun <S> marker(source: S?, primitive: OrientedPoint2D?, styler: ObjectStyler<S?>?): DelegatingPrimitiveGraphic<S?, Point2D?, Graphics2D?>? {
        return DelegatingPrimitiveGraphic(source, primitive, styler, MarkerRenderer.Companion.getInstance())
    }

    fun text(primitive: AnchoredText?): PrimitiveGraphic<AnchoredText?, Graphics2D?>? {
        return PrimitiveGraphic(primitive, Styles.DEFAULT_TEXT_STYLE.copy(), TextRenderer.Companion.getInstance())
    }

    fun text(primitive: AnchoredText?, style: AttributeSet?): PrimitiveGraphic<AnchoredText?, Graphics2D?>? {
        return PrimitiveGraphic(primitive, style, TextRenderer.Companion.getInstance())
    }

    fun <S> text(source: S?, primitive: AnchoredText?, styler: ObjectStyler<S?>?): DelegatingPrimitiveGraphic<S?, AnchoredText?, Graphics2D?>? {
        return DelegatingPrimitiveGraphic(source, primitive, styler, TextRenderer.Companion.getInstance())
    }

    fun image(primitive: AnchoredImage?): PrimitiveGraphic<AnchoredImage?, Graphics2D?>? {
        return PrimitiveGraphic(primitive, AttributeSet.EMPTY, ImageRenderer.Companion.getInstance())
    }

    fun image(x: Double, y: Double, wid: Double, ht: Double, image: Image?, ref: String?): PrimitiveGraphic<AnchoredImage?, Graphics2D?>? {
        return PrimitiveGraphic(AnchoredImage(x, y, wid, ht, image, ref), AttributeSet.EMPTY, ImageRenderer.Companion.getInstance())
    }

    fun icon(icon: AnchoredIcon?): PrimitiveGraphic<AnchoredIcon?, Graphics2D?>? {
        return PrimitiveGraphic(icon, AttributeSet.EMPTY, IconRenderer.Companion.getInstance())
    }

    fun icon(icon: Icon?, x: Double, y: Double): PrimitiveGraphic<AnchoredIcon?, Graphics2D?>? {
        return PrimitiveGraphic(AnchoredIcon(x, y, icon), AttributeSet.EMPTY, IconRenderer.Companion.getInstance())
    }

    fun <S> nodeLink(): DelegatingNodeLinkGraphic<S?, EndpointPair<S?>?, Graphics2D?>? {
        return DelegatingNodeLinkGraphic(MarkerRenderer.Companion.getInstance(), TextRenderer.Companion.getInstance(), PathRenderer.Companion.getInstance())
    } //endregion
}